File: google/main/src/core/query/ExcTerms.java
package main.src.core.query;

import java.util.regex.Pattern;
import java.util.*;

import main.src.core.engine.IndexInterface;
import main.src.core.structures.Token;
import main.src.core.structures.Document;


public class ExcTerms extends Terms{

    private static Pattern pattern = Pattern.compile("\\-(\\w+)");

    public ExcTerms(String expression){
        super();
        this.collect(expression, pattern, 1);
    }

    public HashSet<Document> getResults(IndexInterface index){
        HashSet<Document> results = new HashSet<>();
        for(Token token : this.keys)
            results.addAll(index.getDocumentsOfToken(token));
        return results;
    }
}

File: GoogleSharp/GoogleSharp/src/core/query/ExcTerms.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;
using GoogleSharp.Src.Core.Engine;
using GoogleSharp.Src.Core.Structures;

namespace GoogleSharp.Src.Core.Query
{
    public class ExcTerms : Terms
    {
        private static readonly Regex Pattern = new Regex(@"\-(\w+)", RegexOptions.Compiled | RegexOptions.IgnoreCase);

        public ExcTerms(string expression) : base()
        {
            this.Collect(expression, Pattern, 1);
        }

        public override HashSet<Document> GetResults(IInvertedIndex index)
        {
            var results = new HashSet<Document>();
            this.Tokens.ForEach(t => results.UnionWith(index.GetDocumentsOfToken(t)));
            return results;
        }
    }
}